hyperspy/signal.py

```diff
@@ -4920,6 +4932,84 @@ def T(self):
         """
         return self.transpose()
 
+    def apply_apodization(self, window='hann', hann_order=None, tukey_alpha=0.5, inplace=False):
+        """
+        Apply apodization window to a signal either in place or return a new signal.
+
+        Parameters
+        ----------
+        window : string, optional
+            Select between 'hann', 'hamming', 'tukey'
+            (Default: 'hann')
+        hann_order : None or int, optional
+            Only used if window='hann'
+            If integer n provided hann window of nth order will be used.
+            If None, first order hann window is used.
+            Higher orders result in more homogeneous intensity distribution.
+        tukey_alpha : float
+            Only used if window='tykey'
+            From scipy documentation:
+            Shape parameter of the Tukey window, representing the fraction of
+            the window inside the cosine tapered region. If zero,
+            the Tukey window is equivalent to a rectangular window.
+            If one, the Tukey window is equivalent to a Hann window.
+            (Default: 0.5)
+        inplace : boolean, optional
+            If True apodization applied in place, i.e. signal data will be substituted by the apodized one.
+            (Default: False)
+        Returns
+        -------
+        out : :class:`~hyperspy.signals.BaseSignal or subclasses, optional
+            If 'inplace'=True, returns apodized signal of the same type as self.
+        Examples
+        --------
+        >>> import hyperspy.api as hs
+        >>> holo = hs.datasets.example_signals.object_hologram()
+        >>> holo.apply_apodization('tukey', tukey_alpha=0.1).plot()
+        """
+
+        if window == 'hanning' or window == 'hann':
+            if hann_order:
+                window_function = lambda m: hann_window_nth_order(m, hann_order)
+            else:
+                window_function = lambda m: np.hanning(m)
+        elif window == 'hamming':
+            window_function = lambda m: np.hamming(m)
+        elif window == 'tukey':
+            window_function = lambda m: sp.signal.tukey(m, tukey_alpha)
+        else:
+            raise ValueError('Wrong type parameter value.')
+
+        windows_1d = []
+
+        axes = np.array(self.axes_manager.signal_indices_in_array)
```

Comment:
Would it be more sensible to use the `map` method to do the computation?

Reply:
I don't think it is applicable here... This part constructs nd window function from 1d, which involves creating list of 1d window functions corresponding to the length of each signal axis and than doing outer product. 

The multiplication of each slice with the window is than done at the end (lines 4999 to 5012), and this is actually the part where map functions could have been used... I don't remember the exact reason I've done it manually, perhaps wanted to reduce possible overheads that map introduces with changing dimensionality and etc... Let me know If you would prefer using map there. If needed I can even try benchmarking both options in terms of performance.

Reply:
I suspect this one could have been the reason: #1712

Reply:
Yes, I meant the computation using `self.data` not for the window. I suggested to use `map` because if should be easier to maintain. I don't know if there will be any speed penalty, but normally if it goes through `_map_all` there should not be any significant overheads. But the calculation is fairly simple, so it is not a big deal not to use map.
#1712 is for numpy ufuncs, which operate element wise on the inputs which should not be the case here.

Reply:
> 

> #1712 is for numpy ufuncs, which operate element wise on the inputs which should not be the case here.

It turns out that `np.multiply` is also ufunc:
>np.multiply(*args, **kwargs)
Type:            ufunc
String form:     <ufunc 'multiply'>

So I got the same error...

Reply:
I was more thinking of defining a function which takes an axes parameter and pass this function to map. But that's ok, we can leave it as it is.

Reply:
If I understood you correctly, in this case one would have to recalculate the window each time, which can be quite inefficient.

Reply:
The windows could be a parameter of the function but looking at it now, I agree that using numpy slicing is more simple than going through the `map` hoops! :)
I will wait for CI to finish to check there was not side effect from the last changes and I will merge it.